# Resource files

This document provides an overview for the resource files used in the [WiGiTS](https://github.com/hartwigmedical/hmftools) toolkit.
In file names, `<genome_version>` tags represent either `37` or `38`, referring to GRCh37 or GRCh38 reference genome versions.

## Table of contents
<!-- TOC -->
* [Reference genome](#reference-genome)
* [Common files](#common-files)
  * [Driver config](#driver-config)
  * [Ensembl cache](#ensembl-cache)
* [Variants](#variants)
  * [SNV (single nucleotide variant) annotations](#snv--single-nucleotide-variant--annotations)
  * [SV (structural variant) annotations](#sv--structural-variant--annotations)
  * [CNV (copy number variant) annotations](#cnv--copy-number-variant--annotations)
* [Tool specific](#tool-specific)
  * [REDUX](#redux)
  * [ISOFOX](#isofox)
  * [VIRUSBreakend](#virusbreakend)
  * [Virusinterpreter](#virusinterpreter)
  * [CUPPA](#cuppa)
  * [LILAC](#lilac)
  * [NEO](#neo)
  * [PEACH](#peach)
  * [SIGS](#sigs)
  * [ORANGE](#orange)
<!-- TOC -->

## Reference genome

Genome version prefixes:
- GRCh37: `Homo_sapiens.GRCh37.GATK.illumina`
- GRCh38: `GCA_000001405.15_GRCh38_no_alt_analysis_set`

```shell
<prefix>.fasta       # Genome sequence
<prefix>.fasta.fai   # Sequence index
<prefix>.fasta.dict  # Summary info for each contig
<prefix>.fasta.img   # BWA-MEM2 index image

# BWA-MEM2 indexes
<prefix>.fasta.0123
<prefix>.fasta.amb
<prefix>.fasta.ann
<prefix>.fasta.bwt.2bit.64
<prefix>.fasta.pac

# Indexes used specifically by GRIDSS/VIRUSBreakend
<prefix>.fasta.bwt
<prefix>.fasta.sa
<prefix>.fasta.gridsscache
```

## Common files

### Driver config

```shell
DriverGenePanel.<genome_version>.tsv
```

### Ensembl data cache

Flat file representation Ensembl data generated by HMF [gene-utils](https://github.com/hartwigmedical/hmftools/blob/master/gene-utils/).

```shell
ensembl_trans_splice_data.csv
ensembl_protein_features.csv
ensembl_trans_amino_acids.csv
ensembl_trans_exon_data.csv
ensembl_gene_data.csv
```

## Variants

### SNV (single nucleotide variant) annotations

```shell
# Anotations
gnomad_variants_v<genome_version>.csv.gz
clinvar.<genome_version>.vcf.gz
KnownBlacklist.germline.<genome_version>.bed
KnownBlacklist.germline.<genome_version>.vcf.gz
mappability_150.<genome_version>.bed.gz
SageGermlinePon.1000x.<genome_version>.tsv.

# Region definitions
KnownHotspots.somatic.<genome_version>.vcf.gz
KnownHotspots.germline.<genome_version>.vcf.gz
ActionableCodingPanel.<genome_version>.bed.gz
NA12878_GIAB_highconf_IllFB-IllGATKHC-CG-Ion-Solid_ALLCHROM_v3.2.2_highconf.bed.gz # For GRCh37
HG001_GRCh38_GIAB_highconf_CG-IllFB-IllGATKHC-Ion-10X-SOLID_CHROM1-X_v.3.3.2_highconf_nosomaticdel_noCENorHET7.bed.gz # For GRCh38
CoverageCodingPanel.<genome_version>.bed.gz
```

### SV (structural variant) annotations

```shell
sgl_pon.<genome_version>.bed.gz
sv_pon.<genome_version>.bedpe.gz
known_fusion_data.<genome_version>.csv
known_fusions.<genome_version>.bedpe
sv_prep_blacklist.<genome_version>.bed
repeat_mask_data.<genome_version>.fa.gz
hg38_decoys.fa.img
```

### CNV (copy number variant) annotations

```shell
cohort_germline_del_freq.<genome_version>.csv
AmberGermlineSites.<genome_version>.tsv.gz
DiploidRegions.<genome_version>.bed.gz
GC_profile.1000bp.<genome_version>.cnp
```

## Tool specific

### REDUX

[REDUX](https://github.com/hartwigmedical/hmftools/tree/master/redux) applies various alignment post-processing routines such as duplicate
marking and unmapping of problematic regions for DNA reads.

```shell
unmap_regions.<genome_version>.tsv       # Regions of high depth, repeats or otherwise problematic for mapping
msi_jitter_sites.<genome_version>.tsv.gz # Microsatellite sites used for sample-specific jitter, required for SAGE
```

### ISOFOX

[ISOFOX](https://github.com/hartwigmedical/hmftools/tree/master/isofox) analyses RNA alignment data to quantify
transcripts, identify novel splice junctions, and call fusions.

```shell
isofox.<cohort_size>.gene_distribution.<genome_version>.csv
isofox.<cohort_size>.alt_sj_cohort.<genome_version>.csv
read_<read_length>_exp_counts.<genome_version>.csv
read_<read_length>_exp_gc_ratios.<genome_version>.csv
```

### VIRUSBreakend

[VIRUSBreakend](VIRUSBreakend) detects the presence of oncoviruses and intergration sites in tumor samples. The resource files contain
viral genome and taxonomy data.

```shell
# NCBI RefSeq viral genomes
library/viral/library.fna
library/viral/library.fna.fai
library/viral/library.fna.dict
library/viral/library.fna.masked

# NCBI viral neighbour genomes
library/added/rzW5MDJa1g.fna
library/added/rzW5MDJa1g.fna.fai
library/added/rzW5MDJa1g.fna.dict
library/added/rzW5MDJa1g.fna.masked

# Kraken2 database files
library/hash.k2d
hash.k2d
opts.k2d
taxo.k2d

# Kraken2 mapping of sequences to NCBI taxonomy IDs
seqid2taxid.map

# Table of viral neighbour genomes and their associated RefSeq genome
taxid10239.nbr

# Snapshot of NCBI taxonomy nodes at time of database creation
taxonomy/taxdump.tar.gz
taxonomy/nodes.dmp
```

### Virusinterpreter

[VirusInterpreter](https://github.com/hartwigmedical/hmftools/tree/master/virus-interpreter) performs post-processing for
VIRUSBreakend calls, such as merging of similar virus species into groups.

```shell
virus_blacklisting_db.tsv  # Viruses excluded from reporting in ORANGE for consent reasons
virus_reporting_db.tsv     # Virus groupings
taxonomy_db.tsv            # Taxonomy ID to virus name mappings
```

### CUPPA

[CUPPA](https://github.com/hartwigmedical/hmftools/tree/master/cuppa) predicts tissue of origin for a given tumor sample
using DNA and/or RNA features generated by upstream hmftools components.

```shell
cuppa_classifier.<genome_version>.pickle.gz   # Serialised classifier file
alt_sj.selected_loci.<genome_version>.tsv.gz  # Alternative splice junctions used by the classifier
```

### LILAC

[LILAC](https://github.com/hartwigmedical/hmftools/tree/master/lilac) performs typing of HLA class I genes.

```shell
hla.<genome_version>.bed          # HLA gene regions
hla_ref_nucleotide_sequences.csv  # Nucleotide sequences for each haplotype
hla_ref_aminoacid_sequences.csv   # Coding amino acid sequences for each haplotype
lilac_allele_frequencies.csv      # Population frequencies for each haplotype
```

### NEO

[NEO](https://github.com/hartwigmedical/hmftools/tree/master/neo) identifies neoepitopes from point mutations, small indels and gene fusions,
as well as calculating allele specific neoepitope binding and presentation likelihoods.

```shell
# DNA sequencing
neo_train_expression_dist.csv
neo_train_likelihood.csv
neo_train_exp_likelihood_rand_dist.csv
neo_train_likelihood_rand_dist.csv
neo_train_flank_pos_weight.csv
neo_train_rand_dist.csv
neo_train_pos_weight.csv
neo_train_recognition.csv

# RNA sequencing
isofox.<cohort_size>.transcript_medians.<genome_version>.csv
```

### PEACH

[PEACH](https://github.com/hartwigmedical/hmftools/tree/master/peach) infers the simplest haplotypes supporting variant calls and 
interprets these within a pharmacogenomic context.

```shell
haplotypes.<genome_version>.tsv           # Metadata per gene and per haplotype
haplotype_functions.<genome_version>.tsv  # Functional impact annotations per haplotype
peach_drugs.<genome_version>.tsv          # Relevant drugs per gene
```

### SIGS

[SIGS](https://github.com/hartwigmedical/hmftools/tree/master/sigs) fits defined 
[COSMIC trinucleotide mutational signatures](https://cancer.sanger.ac.uk/signatures/sbs/) to tumor sample data.

```shell
snv_cosmic_signatures.csv  # Mutational signature profiles the original 30 COSMIC signatures
signatures_etiology.tsv    # Mutational signature etiologies
```

### ORANGE

[ORANGE](https://github.com/hartwigmedical/hmftools/tree/master/orange) summaries and integrates key results from
`WiGiTS` components into a single static PDF report.

```shell
cohort_percentiles.tsv  # Cohort stats (e.g. TMB)
cohort_mapping.tsv      # Cancer type name to DOID mapping
doid.json               # Metadata for each DOID
```





